### PR TITLE
Networking fixes

### PR DESCRIPTION
**Pull request by arkpar**
This PR introduces "expired" state for handshakes and sessions. Since all request are processed asynchronously it was possible that a newly created connection receives an event for a previously deleted or about to be deleted connection with the same id. This resulted in some connections being dropped prematurely.

### CODE REVIEW

File: ethcore/res/ethereum/tests [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 3116f85a499ceaf4dfdc46726060fc056e2d7829
+Subproject commit f32954b3ddb5af2dc3dc9ec6d9a28bee848fdf70
```
Comment (arkpar) [inline ethcore/res/ethereum/tests]: Updated test to latest accidentally. They pass anyway

**Comment by gavofyork** [comment]
needs merge.

File: util/src/network/connection.rs [hunk 2]
```diff
@@ -175,13 +175,25 @@ impl Connection {
 		self.socket.peer_addr()
 	}
 
+	pub fn try_clone(&self) -> io::Result<Self> {
```
Comment (gavofyork) [inline util/src/network/connection.rs]: missing docs?

File: util/src/network/connection.rs [hunk 3]
```diff
@@ -265,7 +277,7 @@ impl EncryptedConnection {
 	}
 
 	/// Create an encrypted connection out of the handshake. Consumes a handshake object.
-	pub fn new(mut handshake: Handshake) -> Result<EncryptedConnection, UtilError> {
+	pub fn new(handshake: &mut Handshake) -> Result<EncryptedConnection, UtilError> {
```
Comment (gavofyork) [inline util/src/network/connection.rs]: if it consumes the object, why not pass it by value rather than `&mut`?

File: util/src/network/host.rs [hunk 4]
```diff
@@ -583,9 +583,9 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
 			return;
 		} else if create_session {
 			self.start_session(token, io);
-			io.update_registration(token).unwrap_or_else(|e| debug!(target: "net", "Session registration error: {:?}", e));
+			return;
```
Comment (gavofyork) [inline util/src/network/host.rs]: why don't we `update_registration` any more?

File: util/src/network/host.rs [hunk 5]
```diff
@@ -642,26 +642,25 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
 
 		// turn a handshake into a session
 		let mut sessions = self.sessions.write().unwrap();
-		let mut h = handshakes.remove(token).unwrap();
-		// wait for other threads to stop using it
-		{
-			while Arc::get_mut(&mut h).is_none() {
-				h.lock().ok();
-			}
+		let mut h = handshakes.get_mut(token).unwrap().lock().unwrap();
```
Comment (gavofyork) [inline util/src/network/host.rs]: is this `lock` safe? no chance of anything else later in this scope locking it?

File: util/src/network/host.rs [hunk 6]
```diff
@@ -642,26 +642,25 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
...
+		if h.expired {
+			return;
 		}
-		let h = Arc::try_unwrap(h).ok().unwrap().into_inner().unwrap();
 		let originated = h.originated;
-		let mut session = match Session::new(h, &self.info.read().unwrap()) {
+		let mut session = match Session::new(&mut h, &self.info.read().unwrap()) {
```
Comment (gavofyork) [inline util/src/network/host.rs]: if `h` isn't used again, should be pass-by-move.

File: util/src/network/host.rs [hunk 7]
```diff
@@ -690,26 +689,34 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
 			FIRST_HANDSHAKE ... LAST_HANDSHAKE => {
 				let handshakes = self.handshakes.write().unwrap();
 				if let Some(handshake) = handshakes.get(token).cloned() {
-					failure_id = Some(handshake.lock().unwrap().id().clone());
+					let mut handshake = handshake.lock().unwrap();
```
Comment (gavofyork) [inline util/src/network/host.rs]: is this `lock` safe?

File: util/src/network/host.rs [hunk 8]
```diff
@@ -690,26 +689,34 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
...
 			},
 			FIRST_SESSION ... LAST_SESSION => {
 				let sessions = self.sessions.write().unwrap();
 				if let Some(session) = sessions.get(token).cloned() {
-					let s = session.lock().unwrap();
-					if s.is_ready() {
-						for (p, _) in self.handlers.read().unwrap().iter() {
-							if s.have_capability(p)  {
-								to_disconnect.push(p);
+					let mut s = session.lock().unwrap();
```
Comment (gavofyork) [inline util/src/network/host.rs]: is this `lock` safe?

File: util/src/network/session.rs [hunk 9]
```diff
@@ -109,8 +111,8 @@ const PACKET_USER: u8 = 0x10;
 const PACKET_LAST: u8 = 0x7f;
 
 impl Session {
-	/// Create a new session out of comepleted handshake. Consumes handshake object.
-	pub fn new(h: Handshake, host: &HostInfo) -> Result<Session, UtilError> {
+	/// Create a new session out of comepleted handshake.
+	pub fn new(h: &mut Handshake, host: &HostInfo) -> Result<Session, UtilError> {
```
Comment (gavofyork) [inline util/src/network/session.rs]: consider passing `h` by move.

File: util/src/network/session.rs [hunk 10]
```diff
@@ -169,8 +188,20 @@ impl Session {
 		self.info.capabilities.iter().any(|c| c.protocol == protocol)
 	}
 
+	/// Register the session socket with the event loop
+	pub fn register_socket<Host:Handler<Timeout=Token>>(&self, reg: Token, event_loop: &mut EventLoop<Host>) -> Result<(), UtilError> {
```
Comment (gavofyork) [inline util/src/network/session.rs]: space after `:` and surrounding `=`.

**Comment by gavofyork** [comment]
`self.expired()` in general seems a bit racey. does it matter?

File: util/src/network/host.rs (hunk 4 above)
Comment (arkpar) [inline util/src/network/host.rs]: Before this change session and handshake shared the same connection and it just needed to be updated here. Now sessions gets its own connection and handshake's connection needs to be unregistered instead

File: util/src/network/connection.rs (hunk 3 above)
Comment (arkpar) [inline util/src/network/connection.rs]: It does not anymore, I'll update the comment

File: util/src/network/host.rs (hunk 5 above)
Comment (arkpar) [inline util/src/network/host.rs]: No, there are no other host's methods called in this scope.

File: util/src/network/host.rs (hunk 6 above)
Comment (arkpar) [inline util/src/network/host.rs]: It is used later to deregister the handshake connection. This happens asynchronously in another event loop iteration.

File: util/src/network/host.rs (hunk 7 above)
Comment (arkpar) [inline util/src/network/host.rs]: Not much happens under this lock `io.deregister_stream()` does not lock anything, just sends a message. So I'm pretty sure that it is safe.

File: util/src/network/host.rs (hunk 8 above)
Comment (arkpar) [inline util/src/network/host.rs]: Same here, just some flags checked under this lock

File: util/src/network/session.rs (hunk 9 above)
Comment (arkpar) [inline util/src/network/session.rs]: see comments above

**Comment by arkpar** [comment]
It is always checked with an active lock. It was introduced to make sure that all connection events are ignored from the point when it is decided that connection is no longer needed (`kill_connection`) to the moment it is deregistered and deleted.

**Pull request merged**